table/block_based/block_based_table_builder.cc

```diff
@@ -1630,14 +1644,37 @@ void BlockBasedTableBuilder::EnterUnbuffered() {
   const size_t kSampleBytes = r->compression_opts.zstd_max_train_bytes > 0
                                   ? r->compression_opts.zstd_max_train_bytes
                                   : r->compression_opts.max_dict_bytes;
+
+  // If buffer size is reasonable, we pre-generate a permutation to enforce
+  // uniqueness. This prevents wasting samples on duplicates, which is
+  // particularly likely when not many blocks were buffered.
```

Comment:
Another option to consider vs. random order is some linear ordering that guarantees uniqueness (or almost).

Consider this (classic) option: pick a random starting position, and increment by a large fixed prime, mod max_val. The downside of this is that some max_vals will guarantee poor spread, e.g. if the prime is === 1 (mod number of blocks).

Another option I like (almost uniqueness; less threat of "bad" randomness): treat a w bit number (32 or 64 bits) as a fraction of 0 to max_val (see "FastRange" in our code). Pick a random starting position (a random w-bit number) and each time increment by the w-bit golden ratio. This maximizes the spatial spread of samples[0], which nearly guarantees uniqueness, but also the increased entropy of tracking w bits means there are no "misbehaving" choices of max_val and essentially no chance of bad correlations with underlying data.

[0] You can get slightly more even spread if you use the number of samples in your choice, but that is to the detriment of other properties.

Reply:
This is great, the extra memory needed for storing the vector (or deciding not to store it in certain cases) bothered me a lot. I will probably try the first classic suggestion since it's easier for me to understand and the test cases assume samples are unique. But it'll probably need to be deferred for this to make it into the release...